### PR TITLE
[WIP] serviceaccount: include jti and requester user info

### PR DESCRIPTION
**Pull request by enj**
/kind feature

```release-note
TODO release note
```

```docs
TODO docs
```

---

Start off as cluster admin:

```shell
kubectl auth whoami -o json
```

```json
{
    "kind": "SelfSubjectReview",
    "apiVersion": "authentication.k8s.io/v1",
    "metadata": {
        "creationTimestamp": "2023-08-02T14:40:58Z"
    },
    "status": {
        "userInfo": {
            "username": "system:admin",
            "groups": [
                "system:masters",
                "system:authenticated"
            ]
        }
    }
}
```

Create a token for SA:

```shell
export SA_TOKEN=$(kubectl create token default)
```

Check the claims:

```shell
jwt decode -A RS256 $SA_TOKEN
```

```
Token header
------------
{
  "alg": "RS256",
  "kid": "ZiyTP_HmZMlBFjvqboiTEXJACL0haX7zBxxCsgwiyXU"
}

Token claims
------------
```

```json
{
  "aud": [
    "https://kubernetes.default.svc"
  ],
  "exp": 1690990782,
  "iat": 1690987182,
  "iss": "https://kubernetes.default.svc",
  "jti": "0cf34c52-2ea9-4473-9bd4-0fd3af2a5277",  // NEW
  "kubernetes.io": {
    "namespace": "default",
    "requesterUserInfo": {
      "username": "system:admin"  // NEW
    },
    "serviceaccount": {
      "name": "default",
      "uid": "9a17963b-aa33-43e3-bbb0-551e45120a0d"
    }
  },
  "nbf": 1690987182,
  "sub": "system:serviceaccount:default:default"
}
```

Check the user info of the SA:

```shell
kubectl auth whoami --token=$SA_TOKEN -o json
```

```json
{
    "kind": "SelfSubjectReview",
    "apiVersion": "authentication.k8s.io/v1",
    "metadata": {
        "creationTimestamp": "2023-08-02T14:40:20Z"
    },
    "status": {
        "userInfo": {
            "username": "system:serviceaccount:default:default",
            "uid": "9a17963b-aa33-43e3-bbb0-551e45120a0d",
            "groups": [
                "system:serviceaccounts",
                "system:serviceaccounts:default",
                "system:authenticated"
            ],
            "extra": {
                "authentication.kubernetes.io/jti": [
                    "0cf34c52-2ea9-4473-9bd4-0fd3af2a5277"  // NEW
                ],
                "authentication.kubernetes.io/requester-username": [
                    "system:admin"  // NEW
                ]
            }
        }
    }
}
```

Audit event associated with the token request:

```json
{
  "kind": "Event",
  "apiVersion": "audit.k8s.io/v1",
  "level": "Metadata",
  "auditID": "85eb8b04-dc8e-4f61-9bc8-190285ee3166",
  "stage": "ResponseComplete",
  "requestURI": "/api/v1/namespaces/default/serviceaccounts/default/token",
  "verb": "create",
  "user": {
    "username": "system:admin",
    "groups": [
      "system:masters",
      "system:authenticated"
    ]
  },
  "sourceIPs": [
    "::1"
  ],
  "userAgent": "kubectl/v1.29.0 (darwin/arm64) kubernetes/898a30c",
  "objectRef": {
    "resource": "serviceaccounts",
    "namespace": "default",
    "name": "default",
    "apiVersion": "v1",
    "subresource": "token"
  },
  "responseStatus": {
    "metadata": {},
    "code": 201
  },
  "requestReceivedTimestamp": "2023-08-02T14:39:42.886753Z",
  "stageTimestamp": "2023-08-02T14:39:42.889242Z",
  "annotations": {
    "authentication.kubernetes.io/jti": "0cf34c52-2ea9-4473-9bd4-0fd3af2a5277",  // NEW
    "authorization.k8s.io/decision": "allow",
    "authorization.k8s.io/reason": ""
  }
}
```

Audit event from `selfsubjectreview` via SA token:

```json
{
  "kind": "Event",
  "apiVersion": "audit.k8s.io/v1",
  "level": "Metadata",
  "auditID": "5f3784f0-4fbc-4e01-9e74-dcd788a27cbf",
  "stage": "ResponseComplete",
  "requestURI": "/apis/authentication.k8s.io/v1/selfsubjectreviews",
  "verb": "create",
  "user": {
    "username": "system:serviceaccount:default:default",
    "uid": "9a17963b-aa33-43e3-bbb0-551e45120a0d",
    "groups": [
      "system:serviceaccounts",
      "system:serviceaccounts:default",
      "system:authenticated"
    ],
    "extra": {
      "authentication.kubernetes.io/jti": [
        "0cf34c52-2ea9-4473-9bd4-0fd3af2a5277"  // NEW
      ],
      "authentication.kubernetes.io/requester-username": [
        "system:admin"  // NEW
      ]
    }
  },
  "sourceIPs": [
    "::1"
  ],
  "userAgent": "kubectl/v1.29.0 (darwin/arm64) kubernetes/898a30c",
  "objectRef": {
    "resource": "selfsubjectreviews",
    "apiGroup": "authentication.k8s.io",
    "apiVersion": "v1"
  },
  "responseStatus": {
    "metadata": {},
    "code": 201
  },
  "requestReceivedTimestamp": "2023-08-02T14:40:20.078756Z",
  "stageTimestamp": "2023-08-02T14:40:20.079532Z",
  "annotations": {
    "authorization.k8s.io/decision": "allow",
    "authorization.k8s.io/reason": "RBAC: allowed by ClusterRoleBinding \"system:basic-user\" of ClusterRole \"system:basic-user\" to Group \"system:authenticated\""
  }
}
```

### CODE REVIEW

File: pkg/serviceaccount/claims.go [hunk 1]
```diff
@@ -51,21 +53,33 @@ type kubernetes struct {
 	Pod       *ref             `json:"pod,omitempty"`
 	Secret    *ref             `json:"secret,omitempty"`
 	WarnAfter *jwt.NumericDate `json:"warnafter,omitempty"`
+
+	RequesterUserInfo userInfo `json:"requesterUserInfo,omitempty"`
```
Comment (mikedanese) [inline pkg/serviceaccount/claims.go:57]: Should this be a chain?

File: pkg/serviceaccount/claims.go [hunk 2]
```diff
@@ -51,21 +53,33 @@ type kubernetes struct {
...
 
 type ref struct {
 	Name string `json:"name,omitempty"`
 	UID  string `json:"uid,omitempty"`
 }
 
-func Claims(sa core.ServiceAccount, pod *core.Pod, secret *core.Secret, expirationSeconds, warnafter int64, audience []string) (*jwt.Claims, interface{}) {
+type userInfo struct {
+	// The name that uniquely identifies this user among all active users.
+	Username string `json:"username,omitempty"`
```
Comment (mikedanese) [inline pkg/serviceaccount/claims.go:67]: Can we think through the privacy implications (if any exist) of adding this field and including in things like audit logs?

File: pkg/serviceaccount/claims.go (hunk 1 above)
Comment (munnerz) [reply pkg/serviceaccount/claims.go:57]: A chain of who requested the requesting user's token? I can see why that might be valuable, though I am not totally clear on the use-case for it.

For me at least, a single parent is sufficient to allow me to identify (in my case) which kubelet requested a specific pod's service account token, which allows for a greater level of assurance that a pods token is not being replayed from another host (when the host authenticates to an external service on the pods behalf).

**Comment by munnerz** [comment]
This has been implemented in #120780 and can now be closed 🎉 

/close

**Pull request closed without merging**